gtk: rework Windows and menus

jcollie opened the pull request:
1. Rework the GTK Window code to clean up the if/else spaghetti. This
  _should_ fix issues with older versions of Adwaita getting the
  titlebar and tab bar out of order.
2. Consolidate code for menus into one file and switch to using
  GtkPopupMenus built from GTK Builder XML files. This changes menus
  so that there is one per window and one per surface. This results
  in more memory usage, but more correct behavior. Previously context
  menus would pop up at the wrong location, due to not being attached
  to the correct GTK widget. Using GTK Builder XML files reduces the
  amount of code to create the menus and will make future changes to
  the menu structure easier.
3. Add a "top menu" that can be shown/hidden with a keybind action.
  This will be useful for people that use SSD and thus don't have the
  hamburger menu from the title bar.

![Screenshot From 2025-01-12 14-06-01](https://github.com/user-attachments/assets/6ea230f2-f17b-457b-bcdc-c68e33dc1f8a)
![Screenshot From 2025-01-12 14-06-14](https://github.com/user-attachments/assets/5de8d6cd-d8c6-406f-b935-26e6da1c19f5)

tristan957 commented:
I don't agree with the premise of this PR. I think we should continue to follow patterns established by other GTK/GNOME apps.

unikitty37 commented:
> I don't agree with the premise of this PR. I think we should continue to follow patterns established by other GTK/GNOME apps. "About Ghostty" is not useful in a context menu.

It is if you have `gtk-titlebar` set to `false` and can't access the hamburger menu, though, as there doesn't seem to be any other way to get to it short of adding a keybinding, which is completely excessive :)

Leaving `gtk-titlebar` set to `true` makes Ghostty completely fail to fit in with any WM that _isn't_ Gnome.

tristan957 commented on src/apprt/gtk/ui/menu-window.ui:
```diff
@@ -0,0 +1,92 @@
+<interface domain="com.mitchellh.ghostty">
```
Might be good to add the `<xml>` declaration line.

tristan957 replied:
```
 <?xml version="1.0" encoding="UTF-8">
```

jcollie committed:
gtk: rework Windows and menus

1. Rework the GTK Window code to clean up the if/else spaghetti. This
   _should_ fix issues with older versions of Adwaita getting the
   titlebar and tab bar out of order.
2. Consolidate code for menus into one file and switch to using
   GtkPopupMenus built from GTK Builder XML files. This changes menus
   so that there is one per window and one per surface. This results
   in more memory usage, but more correct behavior. Previously context
   menus would pop up at the wrong location, due to not being attached
   to the correct GTK widget. Using GTK Builder XML files reduces the
   amount of code to create the menus and will make future changes to
   the menu structure easier.
3. Add a "top menu" that can be shown/hidden with a keybind action.
   This will be useful for people that use SSD and thus don't have the
   hamburger menu from the title bar.

mitchellh commented:
There is **a lot** I like in this PR, but there are some fundamental things I don't really like and want more discussion on.

Things I like:

- Menu showing up in the right place
- Cleaned up main menu i.e. supporting all split directions

Things I'm indifferent about (don't care either way):

- Increased memory usage. Its probably small given the benefit.
- GTK builder files.

Things I don't like:

- The context menu and hamburger menu being the same. I don't think they serve the same purpose and long term I don't think it makes sense to always have those match. I also agree with @tristan957 that some times like "About Ghostty" are nonsense in the context menu.
- The context menu is huge (number of items). I'd prefer to make it smaller. To address the non-titlebar people, I like our current approach of nesting a "Menu" submenu. 
- "Top menu" terminology. I don't know what a "top menu" is. Maybe it's common maybe it's not, but Googling at least didn't reveal a lot of usage. To me, there's really just a menu and a context menu. No need to introduce "top".

If we can discuss or address the "don't like" list then I'd be happy to head towards merging this.

jcollie committed:
gtk: menubar update

1. "top menu" => "menubar"
2. Slimmed down context menu.
3. More robust compile time checks of GTK builder UI files.

jcollie committed:
gtk: don't init gtk in builder_check zig because CI doesn't have a display

jcollie commented:
OK, I think that I've addressed all of the feedback so far. "Top bar" was an invention of mine because I couldn't think of anything else, renamed to just "menubar". The context menu is slimmed down. I've also added compile-time checks of
the GTK builder files so it should be much harder to crash Ghostty with an invalid or missing file.

I don't like the idea of embedding the hamburger menu in the context menu, as there's _a lot_ of duplication. It would be much easier to have section(s) of the context menu that get show/hidden based on the state of the titlebar. Not sure what is really "needed" but IMHO that can be bikeshedded later. Since the menus are declared as XML rather than code it's pretty easy to update them.

![Screenshot From 2025-01-24 21-36-12](https://github.com/user-attachments/assets/8df7dfad-80dc-45b1-a030-35c7fc07b46a)

jcollie committed:
gtk: remove stray logging statement

jcollie committed:
gtk: move builder to separate file/struct

jcollie committed:
gtk/adw: add context menu to tab titles to rename tab

jcollie committed:
gtk: work around lack of display in when validating builder files

tristan957 commented:
Thanks for the latest updates. I can't say that I enjoy using the popover menu for the context menu. I still think we should just use the regular rectangular one like all other applications that I can think of.

jcollie commented:
> Thanks for the latest updates. I can't say that I enjoy using the popover menu for the context menu. I still think we should just use the regular rectangular one like all other applications that I can think of.

Is it the arrow that you object to? That's easily removed.

![Screenshot From 2025-01-27 21-15-47](https://github.com/user-attachments/assets/81834532-a149-4679-ad10-19fa6e5f65e7) ![Screenshot From 2025-01-27 21-18-13](https://github.com/user-attachments/assets/7329a3bb-64f5-4168-816e-6e6da307717b)

jcollie commented:
The latest patches add the ability to manually rename tabs. You can restore a manually renamed tab to normal by renaming the tab to a "blank" title. At some point I'd like to add a keybind action to rename the tab (so in theory a tab could be renamed from the keyboard only).

![Screenshot From 2025-01-27 21-22-43](https://github.com/user-attachments/assets/9d7aeb20-42d2-4879-a742-119324a23810)
![Screenshot From 2025-01-27 21-22-55](https://github.com/user-attachments/assets/6e85027f-02c5-444d-b996-cb0de465f1bd)
![Screenshot From 2025-01-27 21-23-01](https://github.com/user-attachments/assets/1f624c5e-0ef2-41dc-965b-9d5794a078e0)

tristan957 commented:
Should this PR maybe be split up for easier reviewing? For instance, fixing the location of the context menu.

jcollie committed:
gtk: don't use arrow on context menu

jcollie commented:
Going to overhaul this once we've removed non-adwaita support.
